### PR TITLE
[BugFix]: Use `single_stream` scheduler with TextGenerationPipeline(s)

### PR DESCRIPTION
**Pull request by rahul-tuli**
Before this PR using text-gen pipelines with deep sparse server would fail when num_streams > 1 is supplied; this is because multi_stream is not supported by text_generation pipelines yet. This PR addresses this failure by always using the the right scheduler(`single_stream`) with text-generation pipelines, and additionally displays appropriate warning log to the user if a non-conforming value is supplied (this used to go through silently before and fail during inference with some sequence lengths)

Step 1: Error resproduction

`server-test-config.yaml`
```yaml
name: opt_model_server_config
num_cores: 2
num_workers: 2
endpoints:
  - task: text_generation
    model: zoo:nlg/text_generation/codegen_mono-350m/pytorch/huggingface/bigpython_bigquery_thepile/base-none
    route: /deterministic/gen
    deterministic: True
```

Test Code:
```python
import requests
r = requests.post("http://0.0.0.0:5543/deterministic/gen", json={"sequences": ["def is_palindrome(word):"]*3})
```

Error Reproduced:

```bash
DeepSparse, Copyright 2021-present / Neuralmagic, Inc. version: 1.6.0.20230821 (458b4ccd) (release) (optimized) (system=avx512, binary=avx512)
OS: Linux office-desktop 5.4.0-156-generic #173-Ubuntu SMP Tue Jul 11 07:25:22 UTC 2023
Arch: x86_64
CPU: GenuineIntel
Vendor: Intel
Cores/sockets/threads: [10, 1, 20]
Available cores/sockets/threads: [10, 1, 20]
L1 cache size data/instruction: 32k/32k
L2 cache size: 1Mb
L3 cache size: 13.75Mb
Thread: 0x7fc8beffa700

Received signal: Segmentation fault (signal addr: 0000000000000010)

Registers:
RAX = 0000000000000001  RBX = 00007fc7b40870a0
RCX = 00007fc7a81d94b0  RDX = 00007fc8beff6430
RDI = 0000000000000000  RSI = 00007fc8beff6350
RBP = 00007fc8beff6350  RSP = 00007fc8beff62c0
 R8 = 00007fc7a81d94b0   R9 = 0000000000000000
R10 = 00007fc8beff6140  R11 = 00007fc7a8000080
R12 = 00007fc7a81d8b88  R13 = 00000000445fd070
R14 = 00007fc8beff6350  R15 = 00007fc8beff6318
RIP = 00007fc93e350b44  EFL = 0000000000010206
CSGSFS = 002b000000000033

 XMM0 = 00007fc7a81d94b000007fc7a81d94b0   XMM1 = 00007fc940a9a38000007fc8b697f040
 XMM2 = 000000000000000000007fc7a81d9340   XMM3 = 00000000000000000000000000000000
 XMM4 = 00000000000000007ff8000000000000   XMM5 = 00007fc9cad92b0000007fc940a9a380
 XMM6 = 00007fc9000000007ff8000000000000   XMM7 = 00007fc9cad92b0000007fc940a9a340
 XMM8 = 00007fc8c63731b000007fc8c63731b0   XMM9 = 00000000000000000000000000000000
XMM10 = 0000000000000000bff0000000000000  XMM11 = 00000000000000000000000000000000
XMM12 = 80feffff80feffff80feffff80feffff  XMM13 = 8000feff8000feff8000feff8000feff
XMM14 = 800000ff800000ff800000ff800000ff  XMM15 = 80ffffff80ffffff80ffffff80ffffff

Backtrace:
 0# 0x00007fc9caf5241f: 
    [0000000f05c3662e0f1f840000000000909066662e0f1f8400000000000f1f40]
    [0048c7c00f0000000f050f1f8000000000f30f1efa4881ec480100004989d064]
 1# 0x00007fc93e350b44: 
    [000000c30f1f840000000000488d87e0000000c30f1f8400000000004883ec08]
    [48837f10007413b8e1a1f5c04839077521488d47184883c408c3ba5201000048]
 2# wand::kv_cache_t::replace_outputs(std::vector<wand::boxed_data_t<5ul>*, std::allocator<wand::boxed_data_t<5ul>*> >&) in /home/rahul/projects/deepsparse/src/deepsparse/avx512/libonnxruntime.so.1.15.1
 3# 0x00007fc93e32e205: 
    [c3384889742468488d7429084939dc75c3488b5424104c89f64c89efe85afcff]
    [ff49837d28000f84bc020000488d742440498d7d184c89f241ff5530488b7c24]
 4# 0x00007fc93d78af8f: 
    [fc3703000f853104000083400801488d8c24800000004c89f24c89fee8702fba]
    [00488bac24880000004885ed7422488b0554fc37034885c00f85b30300008b55]
 5# 0x00007fc93d7d62d2: 
    [00834008014c8b442410488d7c2450488d542470488d8c2490000000e84d48fb]
    [ff4c8b6424684d85e47426488b1d144933034885db0f8593020000418b442408]
 6# 0x00007fc93df1a94d: 
    [380f84f9020000488d4c2440488d5424384c89e7498d71204c8d44246041ff51]
    [38488d65d84c89e05b415c415d415e415f5dc34c8db424d00000004c8d6c2460]
 7# 0x00007fc93dfc933f: 
    [020000000000000f85140b0000498b06488dbc24280100004c89fa4c89f6ff50]
    [10488b8424280100004c8ba4241001000048c784242801000000000000488984]
 8# 0x00007fc93dfbff1c: 
    [0000000041544989f24c89c84889d6498b52084d89c14989c04989fce8c39000]
    [00488b44241049833c24000f94004c89e0415cc341544989fc488b76104889d7]
 9# 0x00007fc93dfcc784: 
    [ff48c78528ffffff00000000488b00488b34d8488b0652488b9518ffffffff50]
    [10584c8b8528ffffff488b8530ffffff5a48c78530ffffff0000000048898528]
10# 0x00007fc93dfc851a: 
    [10c5f97f9dc0fbffff4d85ff0f85e4feffff4c8db5b0fbffff4c89f7e855caff]
    [ffe93cffffff488bbd78fbffffe8a4390000488bbd78fbffffe868390000488b]
11# 0x00007fc93df8dba2: 
    [b550feffffff7520ff7530ff7528ff7510ffb530feffff524c89d250e81d9b03]
    [00488b9d30ffffff4883c4504885db0f84c9040000bf28000000e8ffd8340249]
12# 0x00007fc93df90a92: 
    [78ffffff0100000048c7458000000000c745900000803fc5fa7f4598e8edcbff]
    [ff488b45804883c4404889c74885c07528eb440f1f00488d7f10ba0300000048]
13# 0x00007fc93df910ce: 
    [8b8424f800000050488b4c24504c8b442458488b542448488b742440e821efff]
    [ff4881c4c80000004c89e05b5d415c415d415e415fc34889c5e91da453ff0f1f]
14# 0x00007fc93d6e2c93: 
    [ffff488bb0800600008b80c8010000ffb580faffffffb5c0faffff50e89ce28a]
    [00488b8508fbffff488bbdf8faffff4883c42048c78508fbffff000000004889]
15# 0x00007fc93d67f10a: 
    [ffff6a004c8dad50ffffff524c89ea50574c89df56488bb538feffffe8352a06]
    [00488b8588feffff488bbd80feffff4883c43048c78588feffff000000004889]
16# deepsparse::ort_engine::ort_execute(bool, std::vector<Ort::Value, std::allocator<Ort::Value> > const&, std::vector<Ort::Value, std::allocator<Ort::Value> >&) in /home/rahul/projects/deepsparse/src/deepsparse/avx512/libdeepsparse.so
17# deepsparse::ort_engine::execute_common(bool, std::vector<Ort::Value, std::allocator<Ort::Value> > const&, std::vector<Ort::Value, std::allocator<Ort::Value> >&, std::vector<std::unique_ptr<wand::engine::bench::benchmark_info, std::default_delete<wand::engine::bench::benchmark_info> >, std::allocator<std::unique_ptr<wand::engine::bench::benchmark_info, std::default_delete<wand::engine::bench::benchmark_info> > > >&, std::shared_ptr<wand::kv_cache_t>, bool) in /home/rahul/projects/deepsparse/src/deepsparse/avx512/libdeepsparse.so
18# deepsparse::ort_engine::execute(std::vector<Ort::Value, std::allocator<Ort::Value> > const&, std::vector<std::unique_ptr<wand::engine::bench::benchmark_info, std::default_delete<wand::engine::bench::benchmark_info> >, std::allocator<std::unique_ptr<wand::engine::bench::benchmark_info, std::default_delete<wand::engine::bench::benchmark_info> > > >&, std::shared_ptr<wand::kv_cache_t>) in /home/rahul/projects/deepsparse/src/deepsparse/avx512/libdeepsparse.so
19# 0x00007fc940b77217: 
    [0000004c8da424d0000000488b5424204889ef48894424184d89e04889c141ff]
    [d1488b4c24704c8b6c2478c5f9efc0488b842480000000c5f96fb424f0000000]
20# 0x00007fc940b779eb: 
    [741248833d0b110d00000f855501000083400801488d7c24104889e1e8e4f6ff]
    [ff488b442410488b6c241848c744241000000000498904244885ed0f84a60000]
21# 0x00007fc940b5b686: 
    [eac5fa6f0b4c89f14889ee4c89e7c5f9efc0488b4038c5f97f0c24c5fa7f03ff]
    [d0488b7c24084885ff7405e80af2feff4883c4204c89e05b5d415c415d415ec3]
22# 0x00007fc940bb2852: 
    [3da862090000755683400801488b742460488d7c2408488d5424484c89e941ff]
    [d0488b7c24184885ff7405e83e80f9ff4c8b6424084d85e40f84f5feffff4983]
23# 0x00007fc940b803a9: 
    [2480010000488b78404c89f6e8b6a6fcff4c8dac24c00100004c89ef41ff5424]
    [304889c34c89f74889442478e89650ffff4883fb010f85e71d000048837c2420]
24# PyCFunction_Call in /home/rahul/projects/.venv/bin/python3
25# _PyObject_MakeTpCall in /home/rahul/projects/.venv/bin/python3
26# 0x00000000005ced85: 
    [585b5d415c415d415e415fc349b8ffffffffffffff7f4c89ef4c21c2e81ab0f2]
    [ff49895ef8ebd60f1f40004885c90f85060100004885db0f849e0000004883fb]
27# _PyEval_EvalFrameDefault in /home/rahul/projects/.venv/bin/python3
28# _PyFunction_Vectorcall in /home/rahul/projects/.venv/bin/python3
29# _PyEval_EvalFrameDefault in /home/rahul/projects/.venv/bin/python3

Please email a copy of this stack trace and any additional information to: support@neuralmagic.com
```

# NOTE

On further investigation it was found, that this issue is a direct result of the KVCache object not being thread-safe; this diff does update the TextGenerationPipeline to always use one stream and `single_stream` scheduler but does not address the root cause. But this is still a good to have as it protects against faulty initialization of `TextGenerationPipeline`

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205248270105775

### CODE REVIEW

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 1]
```diff
@@ -718,3 +719,55 @@ def causal_mask_input_present(model_path: str) -> bool:
     def _reset_engines_cache(self):
         self.engine.reset_kv_cache()
         self.multitoken_engine.reset_kv_cache() if self.multitoken_engine else None
+
+    def _update_context(self, **kwargs) -> Context:
```
Comment (dbogunowicz) [inline src/deepsparse/transformers/pipelines/text_generation.py:720]: A bit hestitant to pack more auxiliary functions into `text_generation.py`. Can we somehow import this as a helper func?

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 2]
```diff
@@ -711,3 +712,59 @@ def causal_mask_input_present(model_path: str) -> bool:
             inp.name == "causal_mask"
             for inp in onnx.load(model_path, load_external_data=False).graph.input
         )
+
+    def _reset_engines_cache(self):
```
Comment (dsikka) [inline src/deepsparse/transformers/pipelines/text_generation.py:716]: where is this being used?

File: src/deepsparse/transformers/pipelines/text_generation.py (hunk 1 above)
Comment (dsikka) [reply src/deepsparse/transformers/pipelines/text_generation.py:720]: could probably move it to helpers if we see multi-stream being an issue in other pipelines.

**Review by dsikka — COMMENTED** [review]
LGTM overall. Just a comment and a question

File: src/deepsparse/transformers/pipelines/text_generation.py (hunk 2 above)
Comment (dbogunowicz) [reply src/deepsparse/transformers/pipelines/text_generation.py:716]: This will vanish once @rahul-tuli rebases to the current main

Comment (rahul-tuli) [reply src/deepsparse/transformers/pipelines/text_generation.py:716]: Ooh good catch; I think I mistakenly added this during conflict resolution. Removing!

File: src/deepsparse/transformers/pipelines/text_generation.py (hunk 1 above)
Comment (dbogunowicz) [reply src/deepsparse/transformers/pipelines/text_generation.py:720]: I'd do it now tbh.

Comment (rahul-tuli) [reply src/deepsparse/transformers/pipelines/text_generation.py]: Done!

**Review by dbogunowicz — DISMISSED** [review]
Looking great now

**Review by Satrat — CHANGES_REQUESTED** [review]
Had a few comments, could we also add in a unit test to confirm that requesting multistream and nstreams > 1 correctly sets the values and throws the warning?

File: src/deepsparse/engine.py [hunk 3]
```diff
@@ -169,11 +169,14 @@ class Context(object):
         will try to distribute them evenly across as few sockets as possible.
     :param num_streams: The max number of requests the model can handle
         concurrently.
+    :param scheduler: The kind of scheduler to execute with. Pass None for the default.
```
[the remark this answers is not in the document]
Comment (rahul-tuli) [reply src/deepsparse/engine.py:172]: Oops, good catch the comment is misleading.

File: src/deepsparse/transformers/utils/helpers.py [hunk 4]
```diff
@@ -211,3 +212,57 @@ def create_causal_mask(
...
+                "Only `default`(%s) Scheduler is supported for "
+                "generation pipelines, but got %s."
+                "This `scheduler` argument will be ignored."
+                % (default_scheduler, context.scheduler)
+            )
+        num_cores = context.num_cores
+
+    else:
+        num_streams = kwargs.pop("num_streams", None)
+        num_cores = kwargs.pop("num_cores", None)
```
[the remark this answers is not in the document]
Comment (rahul-tuli) [reply src/deepsparse/transformers/utils/helpers.py:244]: Done!

File: src/deepsparse/transformers/utils/helpers.py [hunk 5]
```diff
@@ -211,3 +212,61 @@ def create_causal_mask(
     causal_mask[:, :, :, :num_zeros] = 0
 
     return causal_mask
+
+
+def update_context_to_single_stream(**kwargs) -> Context:
```
Comment (bfineran) [inline src/deepsparse/transformers/utils/helpers.py:217]: we don't want to create a context unless one already exists

Comment (rahul-tuli) [reply src/deepsparse/transformers/utils/helpers.py:217]: Updating the code, but would like to understand why is this the case?

File: src/deepsparse/engine.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rahul-tuli) [reply src/deepsparse/engine.py]: I asked runtime team and the scheduler should not be updated here

**Review by Satrat — APPROVED** [review]

**Review by dbogunowicz — APPROVED** [review]

**Comment by rahul-tuli** [comment]
Closed in favor of new pipeline refactor

**Pull request closed without merging**